File: 76-minimum-window-substring/minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        int n=s.length();
        int m=t.length();
        unordered_map<char,int> mp;
        int r=0,count=0,minLen=INT_MAX,ind=-1,l=0;
        for(int i=0;i<m;i++) mp[t[i]]++;
        while(r<n)
        {
            if(mp[s[r]]>0) count++;
            mp[s[r]]--;
            while(count==m)
            {
                mp[s[l]]++;
                if(mp[s[l]]>0) count--;
                if(r-l+1<minLen)
                {
                    minLen=r-l+1;
                    ind=l;
                }
                l++;
            }
            r++;
        }
        if(ind==-1) return "";
        string ans=s.substr(ind,minLen);
        return ans;
    }
};

File: 76-minimum-window-substring/minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        int n = s.length();
        int m = t.length();
        int[] hash = new int[256];

        for(char ch : t.toCharArray()) {
            hash[ch]++;
        }

        int l = 0, startIndex = -1;
        int minLen = Integer.MAX_VALUE, count = 0;
        for(int r=0;r<n;r++) {
            char ch = s.charAt(r);
            if(hash[ch] > 0) count++;
            hash[ch]--;

            while(count == m) {
                if(r - l + 1 < minLen) {
                    minLen = r - l + 1;
                    startIndex = l;
                }

                hash[s.charAt(l)]++;
                if(hash[s.charAt(l)] > 0) count--;
                l++;
            }
        }
        if(startIndex == -1) return "";
        else return s.substring(startIndex,startIndex + minLen);
    }
}